File: apps/windows/TeslaSync.App/modals-dialogs/CommandInputDialog.Model.cs
using System.Globalization;
using System.Text.RegularExpressions;
using TeslaSync.App.Core.Notifications;

namespace TeslaSync.App.ModalsDialogs;

/// <summary>
/// The kind of client-side validation a command-input field runs — the native mirror of the web
/// <c>InputField.validation</c> / <c>InputConfig.validation</c> union
/// (<c>'pin' | 'number' | 'decimal' | 'text'</c>, web/src/features/system/commands.ts). It drives both the
/// validation rule (<see cref="CommandInputProjection.Validate"/>) and the field's secret / numeric input
/// affordance (web <c>resolveInputType</c> / <c>resolveInputMode</c>).
/// </summary>
public enum CommandInputValidation
{
    /// <summary>Free text — only the required check applies (web <c>'text'</c> / no validation).</summary>
    Text,

    /// <summary>A 4-digit PIN, masked on entry (web <c>'pin'</c> → <c>type="password"</c>).</summary>
    Pin,

    /// <summary>A whole number within the optional min/max bounds (web <c>'number'</c>).</summary>
    Number,

    /// <summary>A decimal number within the optional min/max bounds (web <c>'decimal'</c>).</summary>
    Real,
}

/// <summary>Wire mapping for <see cref="CommandInputValidation"/> — UI-free so it is asserted headlessly.</summary>
public static class CommandInputValidations
{
    /// <summary>The lower-case web token for <paramref name="validation"/> (web union member).</summary>
    public static string ToToken(CommandInputValidation validation) => validation switch
    {
        CommandInputValidation.Pin => "pin",
        CommandInputValidation.Number => "number",
        CommandInputValidation.Real => "decimal",
        _ => "text",
    };

    /// <summary>
    /// Parse a web validation token to a <see cref="CommandInputValidation"/>; an unknown / null token maps to
    /// <see cref="CommandInputValidation.Text"/> (web parity — the <c>default</c> branch performs no validation).
    /// </summary>
    public static CommandInputValidation FromToken(string? token) => token switch
    {
        "pin" => CommandInputValidation.Pin,
        "number" => CommandInputValidation.Number,
        "decimal" => CommandInputValidation.Real,
        _ => CommandInputValidation.Text,
    };
}

/// <summary>
/// One field of a multi-field command-input form — the native mirror of the web <c>InputField</c>
/// (web/src/features/system/commands.ts). <see cref="LabelKey"/> / <see cref="LabelFallback"/> resolve the
/// field label through the i18n facade (web <c>t(field.labelKey, field.labelFallback)</c>); the input
/// type / mode are derived from <see cref="Validation"/> (web ignores <c>field.type</c> and uses
/// <c>resolveInputType(field.validation)</c>).
/// </summary>
public sealed record CommandInputField(
    string Name,
    string LabelKey,
    string LabelFallback,
    string? Hint = null,
    CommandInputValidation Validation = CommandInputValidation.Text,
    double? Min = null,
    double? Max = null);

/// <summary>
/// The command-input form the dialog binds — the native projection of a web <c>CommandDef</c> plus its
/// <c>inputConfig</c> (web/src/features/system/commands.ts). It carries the header copy (title + prompt + the
/// optional single-field sub-label + a Segoe Fluent header glyph standing in for the web lucide
/// <c>def.icon</c>) and the input shape: either a list of <see cref="Fields"/> (the multi-field branch, e.g.
/// HomeLink lat/lon) or a single parameter keyed by <see cref="ParamName"/> with its <see cref="Validation"/>
/// / <see cref="Min"/> / <see cref="Max"/> and an initial value (web <c>getDefaultValue?.({ vehicle }) ??
/// defaultValue ?? ''</c> via <see cref="ResolveDefaultValue"/> / <see cref="DefaultValue"/>). It is a plain,
/// WinUI-free data carrier so the projection + view-model are asserted headlessly.
/// </summary>
public sealed class CommandInputForm
{
    /// <summary>The dialog title i18n key (web <c>def.labelKey</c>).</summary>
    public required string TitleKey { get; init; }

    /// <summary>The dialog title English fallback (web <c>def.labelFallback</c>).</summary>
    public required string TitleFallback { get; init; }

    /// <summary>The single-field sub-label i18n key (web <c>def.sublabelKey</c>); null hides the label.</summary>
    public string? SubtitleKey { get; init; }

    /// <summary>The single-field sub-label fallback (web <c>def.sublabelFallback</c>); null hides the label.</summary>
    public string? SubtitleFallback { get; init; }

    /// <summary>A Segoe Fluent glyph standing in for the web lucide <c>def.icon</c>; empty hides the header icon.</summary>
    public string IconGlyph { get; init; } = string.Empty;

    /// <summary>The prompt i18n key shown under the title (web <c>ic.promptKey</c>).</summary>
    public required string PromptKey { get; init; }

    /// <summary>The prompt English fallback (web <c>ic.promptFallback</c>).</summary>
    public required string PromptFallback { get; init; }

    /// <summary>The single-field parameter name (web <c>ic.paramName</c>).</summary>
    public required string ParamName { get; init; }

    /// <summary>The single-field default value used as the input hint (web <c>ic.defaultValue</c>).</summary>
    public string? DefaultValue { get; init; }

    /// <summary>The single-field validation rule (web <c>ic.validation</c>).</summary>
    public CommandInputValidation Validation { get; init; } = CommandInputValidation.Text;

    /// <summary>The single-field lower bound for number/decimal validation (web <c>ic.min</c>).</summary>
    public double? Min { get; init; }

    /// <summary>The single-field upper bound for number/decimal validation (web <c>ic.max</c>).</summary>
    public double? Max { get; init; }

    /// <summary>
    /// The multi-field list (web <c>ic.fields</c>); when non-null the dialog renders one input per field and
    /// the single-field <see cref="ParamName"/> / <see cref="Validation"/> / bounds are not used.
    /// </summary>
    public IReadOnlyList<CommandInputField>? Fields { get; init; }

    /// <summary>
    /// Resolves the single-field initial value from the active vehicle's display name — the native analogue of
    /// the web <c>ic.getDefaultValue?.({ vehicle })</c> closure (the only real use is "Rename", which seeds the
    /// field with the current name). When set it takes precedence over <see cref="DefaultValue"/> (web parity).
    /// </summary>
    public Func<string?, string>? ResolveDefaultValue { get; init; }
}

/// <summary>
/// The submission the dialog emits — the native analogue of the web <c>onSubmit(values)</c> argument
/// (web/src/features/system/components/CommandInputDialog.tsx). <see cref="Values"/> maps each field name to
/// its raw entered value (a single <c>{ paramName: value }</c> entry in the single-field branch, or one entry
/// per <c>ic.fields</c> in the multi-field branch); the parent owns turning it into command parameters
/// (web <c>buildParams</c>).
/// </summary>
public sealed record CommandInputSubmission(IReadOnlyDictionary<string, string> Values);

/// <summary>
/// Canonical slug + the static i18n keys the <c>CommandInputDialog</c> surface owns — the native mirror of
/// <c>web/src/features/system/components/CommandInputDialog.tsx</c>. The dynamic copy (title, prompt, field
/// labels, sub-label) is keyed per command in the bound <see cref="CommandInputForm"/> and resolved through the
/// facade exactly like the web <c>t(def.labelKey, …)</c> calls; this registry keys the surface-owned literals:
/// the Cancel / Send buttons (web <c>t('common.cancel', 'Cancel')</c> / <c>t('common.send', 'Send')</c>) and
/// the validation messages the web component returns as bare strings (so the native view + view-model never
/// embed an English literal). UI-free so every key, fallback and the <c>{{min}}</c> / <c>{{max}}</c>
/// interpolation are asserted headlessly.
/// </summary>
public static class CommandInputRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event (P1/S11 contract).</summary>
    public const string Slug = "CommandInputDialog";

    /// <summary>Cancel button label (web <c>t('common.cancel', 'Cancel')</c>).</summary>
    public static string CancelLabel(ILocalizer localizer) =>
        Require(localizer).GetString("common.cancel", "Cancel");

    /// <summary>Submit button label (web <c>t('common.send', 'Send')</c>).</summary>
    public static string SubmitLabel(ILocalizer localizer) =>
        Require(localizer).GetString("common.send", "Send");

    /// <summary>Required-field message (web <c>validateField</c> returns the literal <c>'Required'</c>).</summary>
    public static string RequiredMessage(ILocalizer localizer) =>
        Require(localizer).GetString("commands.input.required", "Required");

    /// <summary>PIN-format message (web <c>'Enter a 4-digit PIN'</c>).</summary>
    public static string PinMessage(ILocalizer localizer) =>
        Require(localizer).GetString("commands.input.pinFormat", "Enter a 4-digit PIN");

    /// <summary>Whole-number message (web <c>'Enter a whole number'</c>).</summary>
    public static string WholeNumberMessage(ILocalizer localizer) =>
        Require(localizer).GetString("commands.input.wholeNumber", "Enter a whole number");

    /// <summary>Decimal-number message (web <c>'Enter a valid number'</c>).</summary>
    public static string DecimalMessage(ILocalizer localizer) =>
        Require(localizer).GetString("commands.input.validNumber", "Enter a valid number");

    /// <summary>
    /// Minimum-bound message with the bound interpolated — the native analogue of web
    /// <c>`Minimum: ${min}`</c>. Resolves the keyed template then substitutes the i18next <c>{{min}}</c> token
    /// so a translated catalog string keeps the substitution point (the FeedbackModal interpolation convention).
    /// </summary>
    public static string MinimumMessage(ILocalizer localizer, double min)
    {
        string template = Require(localizer).GetString("commands.input.minimum", "Minimum: {{min}}");
        return template.Replace("{{min}}", FormatBound(min), StringComparison.Ordinal);
    }

    /// <summary>
    /// Maximum-bound message with the bound interpolated — the native analogue of web <c>`Maximum: ${max}`</c>
    /// (see <see cref="MinimumMessage"/> for the interpolation convention).
    /// </summary>
    public static string MaximumMessage(ILocalizer localizer, double max)
    {
        string template = Require(localizer).GetString("commands.input.maximum", "Maximum: {{max}}");
        return template.Replace("{{max}}", FormatBound(max), StringComparison.Ordinal);
    }

    /// <summary>
    /// Format a numeric bound the way the web template literal does (JS <c>${min}</c>): a whole value prints
    /// without a fractional part (<c>50</c>), a fractional value keeps it (<c>15.5</c>), invariant culture.
    /// </summary>
    public static string FormatBound(double bound) =>
        bound.ToString("0.################", CultureInfo.InvariantCulture);

    private static ILocalizer Require(ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(localizer);
        return localizer;
    }
}

/// <summary>
/// Pure projections for the <c>CommandInputDialog</c> surface — the native analogue of the web component's
/// <c>validateField</c> rule, its <c>buildInitialValues</c> seeding and its <c>isValid</c> submit gate
/// (web/src/features/system/components/CommandInputDialog.tsx). Every message flows through the i18n facade so
/// the rule is asserted headlessly and the view-model never recomputes it inline. The number / decimal parsers
/// reproduce JavaScript <c>parseInt(v, 10)</c> + <c>String(n) === v</c> and <c>parseFloat</c> semantics so the
/// accept / reject boundary matches the web byte-for-byte.
/// </summary>
public static partial class CommandInputProjection
{
    /// <summary>
    /// Validate a field value — the native analogue of web <c>validateField(value, validation, min, max)</c>.
    /// Returns the localized error message, or null when the value is acceptable. A blank value always fails the
    /// required check first (web <c>if (!trimmed) return 'Required'</c>); otherwise the rule depends on
    /// <paramref name="validation"/>.
    /// </summary>
    public static string? Validate(
        string? value,
        CommandInputValidation validation,
        double? min,
        double? max,
        ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(localizer);

        string trimmed = (value ?? string.Empty).Trim();
        if (trimmed.Length == 0)
        {
            return CommandInputRegistration.RequiredMessage(localizer);
        }

        switch (validation)
        {
            case CommandInputValidation.Pin:
                return PinRegex().IsMatch(trimmed) ? null : CommandInputRegistration.PinMessage(localizer);

            case CommandInputValidation.Number:
                if (!TryParseJsInt(trimmed, out long whole))
                {
                    return CommandInputRegistration.WholeNumberMessage(localizer);
                }

                return CheckBounds(whole, min, max, localizer);

            case CommandInputValidation.Real:
                if (!TryParseJsFloat(trimmed, out double real))
                {
                    return CommandInputRegistration.DecimalMessage(localizer);
                }

                return CheckBounds(real, min, max, localizer);

            default:
                return null;
        }
    }

    /// <summary>
    /// Seed the form's values — the native analogue of web <c>buildInitialValues()</c>. The multi-field branch
    /// returns one empty entry per field (web <c>for (const f of fields) vals[f.name] = ''</c>); the
    /// single-field branch returns one entry keyed by <see cref="CommandInputForm.ParamName"/> whose value is
    /// <c>ResolveDefaultValue(vehicleDisplayName)</c> when supplied, else <see cref="CommandInputForm.DefaultValue"/>,
    /// else the empty string (web <c>getDefaultValue ? getDefaultValue({ vehicle }) : defaultValue ?? ''</c>).
    /// </summary>
    public static Dictionary<string, string> BuildInitialValues(CommandInputForm form, string? vehicleDisplayName)
    {
        ArgumentNullException.ThrowIfNull(form);

        if (form.Fields is { } fields)
        {
            var values = new Dictionary<string, string>(fields.Count, StringComparer.Ordinal);
            foreach (var field in fields)
            {
                values[field.Name] = string.Empty;
            }

            return values;
        }

        string seed = form.ResolveDefaultValue is { } resolve
            ? resolve(vehicleDisplayName) ?? string.Empty
            : form.DefaultValue ?? string.Empty;

        return new Dictionary<string, string>(1, StringComparer.Ordinal) { [form.ParamName] = seed };
    }

    /// <summary>
    /// The live submit gate — the native analogue of web <c>isValid()</c>. Returns true when every field passes
    /// <see cref="Validate"/> (the multi-field branch validates each <c>ic.fields</c> entry; the single-field
    /// branch validates <see cref="CommandInputForm.ParamName"/>). Missing entries are treated as empty so an
    /// absent required value keeps the gate closed.
    /// </summary>
    public static bool IsValid(
        CommandInputForm form,
        IReadOnlyDictionary<string, string> values,
        ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(form);
        ArgumentNullException.ThrowIfNull(values);
        ArgumentNullException.ThrowIfNull(localizer);

        if (form.Fields is { } fields)
        {
            foreach (var field in fields)
            {
                if (Validate(Lookup(values, field.Name), field.Validation, field.Min, field.Max, localizer) is not null)
                {
                    return false;
                }
            }

            return true;
        }

        return Validate(Lookup(values, form.ParamName), form.Validation, form.Min, form.Max, localizer) is null;
    }

    /// <summary>
    /// Reproduces JavaScript <c>parseInt(v, 10)</c> followed by the web <c>String(num) !== v</c> guard: succeeds
    /// only when <paramref name="trimmed"/> is the canonical base-10 representation of an integer (rejecting
    /// leading zeros, a leading <c>+</c>, decimals, exponents and trailing text), returning the parsed value.
    /// </summary>
    public static bool TryParseJsInt(string trimmed, out long value)
    {
        value = 0;
        if (!long.TryParse(trimmed, NumberStyles.AllowLeadingSign, CultureInfo.InvariantCulture, out value))
        {
            return false;
        }

        return string.Equals(value.ToString(CultureInfo.InvariantCulture), trimmed, StringComparison.Ordinal);
    }

    /// <summary>
    /// Reproduces JavaScript <c>parseFloat(v)</c>: parses the leading numeric token (optional sign, digits,
    /// optional fraction, optional exponent), ignoring any trailing text (so <c>"1.5abc"</c> yields
    /// <c>1.5</c>), and fails (NaN) only when no number leads the string.
    /// </summary>
    public static bool TryParseJsFloat(string trimmed, out double value)
    {
        value = double.NaN;
        var match = LeadingFloatRegex().Match(trimmed);
        if (!match.Success || match.Index != 0)
        {
            return false;
        }

        string token = match.Value;
        if (token.EndsWith('.'))
        {
            token = token[..^1];
        }

        if (token.Length == 0 || token == "+" || token == "-")
        {
            return false;
        }

        return double.TryParse(token, NumberStyles.Float, CultureInfo.InvariantCulture, out value);
    }

    private static string? CheckBounds(double value, double? min, double? max, ILocalizer localizer)
    {
        if (min is { } lo && value < lo)
        {
            return CommandInputRegistration.MinimumMessage(localizer, lo);
        }

        if (max is { } hi && value > hi)
        {
            return CommandInputRegistration.MaximumMessage(localizer, hi);
        }

        return null;
    }

    private static string Lookup(IReadOnlyDictionary<string, string> values, string name) =>
        values.TryGetValue(name, out var value) ? value : string.Empty;

    [GeneratedRegex(@"^\d{4}$")]
    private static partial Regex PinRegex();

    [GeneratedRegex(@"^[+-]?(\d+\.?\d*|\.\d+)([eE][+-]?\d+)?")]
    private static partial Regex LeadingFloatRegex();
}

/// <summary>
/// PII-safe diagnostics for the <c>CommandInputDialog</c> surface (P1/S11 diagnostics contract). Records only
/// operational counters with the surface slug — never the entered PIN / value(s) — so a diagnostics line can
/// never leak a command parameter. Thread-safe.
/// </summary>
public sealed class CommandInputDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;
    private long _submitted;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    public CommandInputDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Number of submissions emitted from this surface.</summary>
    public long Submitted => Interlocked.Read(ref _submitted);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=CommandInputDialog</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke(string.Create(
            CultureInfo.InvariantCulture, $"view.opened slug={CommandInputRegistration.Slug}"));
    }

    /// <summary>Record that a submission was emitted (the entered values are never logged).</summary>
    public void RecordSubmitted()
    {
        Interlocked.Increment(ref _submitted);
        _sink?.Invoke(string.Create(
            CultureInfo.InvariantCulture, $"command.input.submitted slug={CommandInputRegistration.Slug}"));
    }
}

File: apps/android/app/src/main/java/com/teslasync/modals-dialogs/CommandInputDialog/CommandInputDialogModel.kt
// Pure, framework-free model + projection for the CommandInputDialog modal/dialog surface — the native analogue
// of everything the web component derives before it returns JSX
// (web/src/features/system/components/CommandInputDialog.tsx). No Compose, no Android, no HTTP: every declaration
// here is exercised off-device by the :android:testReleaseUnitTest gate, so the composable stays a thin render
// layer over these pure functions.
//
// The web component is the vehicle-command parameter prompt. It is a *controlled* dialog whose only data
// dependency is `useTranslation` (i18n, P1/S10) — it binds no fetch and owns no store. Its command spec (`def` /
// `inputConfig`) and its submit/close handlers are caller-supplied props, and the `loading` flag reflects the
// OWNING command page's mutation, not this dialog's. So (exactly like the sibling ConfirmDialog surface) the
// cache-then-network lifecycle (loading / empty / error / stale / offline) belongs to the owner that decides to
// raise the prompt, not here; modelling those phases would invent behaviour the web spec does not have (drift).
// The branches the web source actually defines are the complete state set this surface renders, and each is
// projected here:
//   1. the per-field validation (web `validateField`) — Required on empty, the 4-digit-PIN / whole-number /
//      decimal rules, and the min / max bound checks — surfaced as a typed [FieldError] so the localized message
//      is resolved at the Compose boundary (P1/S10), never as a literal here,
//   2. the initial form values (web `buildInitialValues`) — empty for every multi-field, the resolved default
//      for the single-param case (the caller bakes `getDefaultValue({ vehicle })` / `defaultValue` into
//      [CommandInputField.initialValue]),
//   3. the whole-form validity gate (web `isValid`) that drives the disabled Send button,
//   4. the on-submit pass (web `handleSubmit`) that validates + touches every field at once,
//   5. the keyboard / masking resolution (web `resolveInputType` / `resolveInputMode`) — PIN masks + numeric,
//      number numeric, decimal decimal, text plain.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/modals-dialogs/CommandInputDialog — the P3 prompt's allowed-files path) cannot form a valid
// Kotlin package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package
// intentionally diverges from the path — exactly as the sibling modal surfaces do. `MatchingDeclarationName` is
// suppressed for the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.modalsdialogs.commandinputdialog

import io.teslasync.shared.core.diagnostics.Logger

/**
 * The validation vocabulary a command parameter can declare — the native union of the web
 * `validation?: 'pin' | 'number' | 'decimal' | 'text'` (with [None] standing in for the web `undefined`). It
 * selects both the [CommandInputDialogProjection.validate] rule and the [KeyboardKind] / masking the field uses.
 */
enum class FieldValidation {
    /** No declared rule (web `undefined`): any non-empty value passes. */
    None,

    /** Free text (web `'text'`): any non-empty value passes; plain keyboard. */
    Text,

    /** A 4-digit PIN (web `'pin'`): masked numeric input matching `^\d{4}$`. */
    Pin,

    /** A whole number (web `'number'`): canonical base-10 integer, optionally bounded by min / max. */
    Number,

    /** A decimal number (web `'decimal'`): a parseable float, optionally bounded by min / max. */
    Decimal,
}

/**
 * The typed result of validating one field — the native analogue of the English string `validateField` returns
 * inline in the web source. Kept as a type (not a message) so the composable resolves the localized copy at the
 * Compose boundary (P1/S10); the bounded cases carry the numeric bound the message interpolates (web
 * `Minimum: ${min}` / `Maximum: ${max}`). A `null` result means the field is valid.
 */
sealed interface FieldError {
    /** The value was empty after trimming (web `if (!trimmed) return 'Required'`). */
    data object Required : FieldError

    /** A PIN field whose value is not exactly four digits (web `'Enter a 4-digit PIN'`). */
    data object Pin : FieldError

    /** A number field whose value is not a canonical whole number (web `'Enter a whole number'`). */
    data object WholeNumber : FieldError

    /** A decimal field whose value is not a parseable number (web `'Enter a valid number'`). */
    data object ValidNumber : FieldError

    /** The numeric value is below the declared minimum (web `Minimum: ${min}`); [bound] is that minimum. */
    data class Min(
        val bound: Double,
    ) : FieldError

    /** The numeric value is above the declared maximum (web `Maximum: ${max}`); [bound] is that maximum. */
    data class Max(
        val bound: Double,
    ) : FieldError
}

/**
 * The keyboard flavour a field requests — the native fusion of the web `resolveInputMode` (numeric / decimal /
 * text) and `resolveInputType` (password for PIN). The Compose boundary maps each to a `KeyboardType` plus, for
 * [NumericPassword], a masking `VisualTransformation`.
 */
enum class KeyboardKind {
    /** Plain text keyboard (web `inputMode='text'`, `type='text'`). */
    Text,

    /** Numeric keyboard (web `inputMode='numeric'`, `type='text'`). */
    Numeric,

    /** Decimal keyboard (web `inputMode='decimal'`, `type='text'`). */
    Decimal,

    /** Masked numeric keyboard for PINs (web `inputMode='numeric'`, `type='password'`). */
    NumericPassword,
}

/**
 * One editable field the dialog renders — the native, already-localized projection of a web `InputField`
 * (multi-field case) or the single `inputConfig` param. [label] / [hint] are caller-localized (web
 * `t(field.labelKey, …)` / `def.sublabel`), so the pure layer never holds an i18n key. [initialValue] mirrors the
 * web `buildInitialValues` seed: empty for the multi-field case, the resolved default for the single-param case.
 *
 * @property name the wire param name (web `field.name` / `inputConfig.paramName`); the [onSubmit] map key.
 * @property label the caller-localized field label (web `t(field.labelKey, field.labelFallback)` / the single
 *   param's `def.sublabel`); `null` renders an unlabelled field (web single param with no `sublabelFallback`).
 * @property hint the caller-supplied example text shown in an empty field (web input example text /
 *   `inputConfig.defaultValue`).
 * @property validation the rule this field enforces (web `field.validation` / `inputConfig.validation`).
 * @property min the inclusive lower bound for number / decimal fields (web `min`); `null` when unbounded.
 * @property max the inclusive upper bound for number / decimal fields (web `max`); `null` when unbounded.
 * @property initialValue the value the field opens with (web `buildInitialValues`): `""` for multi-field, the
 *   resolved default for the single-param case.
 */
data class CommandInputField(
    val name: String,
    val label: String?,
    val hint: String?,
    val validation: FieldValidation = FieldValidation.None,
    val min: Double? = null,
    val max: Double? = null,
    val initialValue: String = "",
)

/**
 * The fully-localized command-prompt spec the dialog renders — the native projection of the web `def` +
 * `inputConfig` the component reads. [title] / [prompt] are caller-localized (web `t(def.labelKey, …)` /
 * `t(inputConfig.promptKey, …)`). [fields] is always non-empty: the web's single-param branch is modelled as a
 * one-element list (whose sole field carries the `paramName`, the resolved default, and the param validation), so
 * the two web render branches unify into one list with no behaviour loss — `onSubmit` still emits exactly the web
 * `{ [paramName]: value }` / `{ [field.name]: value, … }` shape.
 *
 * @property title the dialog title (web `t(def.labelKey, def.labelFallback)`); caller-localized.
 * @property prompt the prompt line under the title (web `t(inputConfig.promptKey, inputConfig.promptFallback)`).
 * @property fields the ordered fields to render (web `inputConfig.fields` or the single `paramName`); non-empty.
 */
data class CommandInputSpec(
    val title: String,
    val prompt: String,
    val fields: List<CommandInputField>,
)

/**
 * Pure projection from the dialog's inputs to its render + validation decisions — a 1:1 port of the derivations
 * the web component performs (the `validateField` switch, `buildInitialValues`, `isValid`, the `handleSubmit`
 * touch-all pass, and the `resolveInputType` / `resolveInputMode` keyboard mapping). No Compose, no side effects,
 * so it is fully covered by the off-device unit gate.
 */
object CommandInputDialogProjection {
    private val pinRegex = Regex("^\\d{4}$")
    private val integerPrefixRegex = Regex("^[+-]?\\d+")
    private val floatPrefixRegex = Regex("^[+-]?(?:\\d+\\.?\\d*|\\.\\d+)(?:[eE][+-]?\\d+)?")

    /**
     * The initial value map the form opens with — the web `buildInitialValues`. Each field seeds its own
     * [CommandInputField.initialValue] (`""` for multi-field, the resolved default for the single-param case).
     */
    fun initialValues(spec: CommandInputSpec): Map<String, String> = spec.fields.associate { it.name to it.initialValue }

    /**
     * Validates [field]'s current [value] — the web `validateField(value, field.validation, field.min,
     * field.max)`. Returns `null` when the field is valid.
     */
    fun validate(
        field: CommandInputField,
        value: String,
    ): FieldError? = validate(value, field.validation, field.min, field.max)

    /**
     * The core single-field validator — a faithful port of the web `validateField`. An empty (trimmed) value is
     * always [FieldError.Required]; otherwise the [validation] rule decides. Text / none always pass once
     * non-empty.
     */
    fun validate(
        value: String,
        validation: FieldValidation,
        min: Double?,
        max: Double?,
    ): FieldError? {
        val trimmed = value.trim()
        if (trimmed.isEmpty()) return FieldError.Required
        return when (validation) {
            FieldValidation.Pin -> if (pinRegex.matches(trimmed)) null else FieldError.Pin
            FieldValidation.Number -> validateWholeNumber(trimmed, min, max)
            FieldValidation.Decimal -> validateDecimal(trimmed, min, max)
            FieldValidation.Text, FieldValidation.None -> null
        }
    }

    /**
     * Whether the whole form passes — the web `isValid`. Drives the disabled Send button: every field must
     * validate to `null` against its current value (missing keys treated as empty, i.e. [FieldError.Required]).
     */
    fun isValid(
        spec: CommandInputSpec,
        values: Map<String, String>,
    ): Boolean = spec.fields.all { validate(it, values[it.name].orEmpty()) == null }

    /**
     * The validate-and-touch-all pass the web `handleSubmit` runs before deciding to submit — returns the error
     * for every field (web builds `newErrors` + `newTouched` for all fields). The composable marks every field
     * touched and submits only when the whole map is error-free (web `if (valid) onSubmit(values)`).
     */
    fun errorsOnSubmit(
        spec: CommandInputSpec,
        values: Map<String, String>,
    ): Map<String, FieldError?> = spec.fields.associate { it.name to validate(it, values[it.name].orEmpty()) }

    /**
     * The keyboard flavour for a [validation] — the web `resolveInputMode` fused with `resolveInputType`. PIN is
     * masked numeric; number is numeric; decimal is decimal; text / none are plain.
     */
    fun keyboardKind(validation: FieldValidation): KeyboardKind =
        when (validation) {
            FieldValidation.Pin -> KeyboardKind.NumericPassword
            FieldValidation.Number -> KeyboardKind.Numeric
            FieldValidation.Decimal -> KeyboardKind.Decimal
            FieldValidation.Text, FieldValidation.None -> KeyboardKind.Text
        }

    /** Whether a [validation]'s input should be masked — the web `resolveInputType === 'password'` (PIN only). */
    fun isMasked(validation: FieldValidation): Boolean = validation == FieldValidation.Pin

    /**
     * Formats a numeric [bound] the way the web interpolates `${min}` / `${max}` — an integral value renders with
     * no fractional part (`50`, `30`), a genuinely fractional one renders normally (`0.5`). Mirrors JS
     * `String(Number)` for the bounds the command catalog actually declares.
     */
    fun formatBound(bound: Double): String =
        when {
            bound.isFinite() && bound % 1.0 == 0.0 -> bound.toLong().toString()
            else -> bound.toString()
        }

    private fun validateWholeNumber(
        trimmed: String,
        min: Double?,
        max: Double?,
    ): FieldError? {
        val parsed = jsParseInt(trimmed)
        // web: `isNaN(num) || String(num) !== trimmed` — reject non-canonical integers ("01", "+5", "1.5", "1e3").
        if (parsed == null || parsed.toString() != trimmed) return FieldError.WholeNumber
        // Compare the parsed integer directly against the bounds (web compares `num`); no float widening needed.
        return when {
            min != null && parsed < min -> FieldError.Min(min)
            max != null && parsed > max -> FieldError.Max(max)
            else -> null
        }
    }

    private fun validateDecimal(
        trimmed: String,
        min: Double?,
        max: Double?,
    ): FieldError? {
        val parsed = jsParseFloat(trimmed) ?: return FieldError.ValidNumber
        return checkBounds(parsed, min, max)
    }

    private fun checkBounds(
        num: Double,
        min: Double?,
        max: Double?,
    ): FieldError? =
        when {
            min != null && num < min -> FieldError.Min(min)
            max != null && num > max -> FieldError.Max(max)
            else -> null
        }

    /**
     * The leading-integer parse the web gets from `parseInt(trimmed, 10)`: an optional sign then base-10 digits,
     * stopping at the first non-digit. Returns `null` for the JS `NaN` case (no leading digits). The caller
     * re-stringifies the result to reproduce the web `String(num) !== trimmed` canonical-form check.
     */
    private fun jsParseInt(trimmed: String): Long? = integerPrefixRegex.find(trimmed)?.value?.toLongOrNull()

    /**
     * The leading-float parse the web gets from `parseFloat(trimmed)`: an optional sign, an integer / fractional
     * mantissa, and an optional exponent, stopping at the first character that cannot extend the number (so
     * "1.5abc" yields 1.5, matching JS). Returns `null` for the JS `NaN` case (no leading number).
     */
    private fun jsParseFloat(trimmed: String): Double? {
        val match = floatPrefixRegex.find(trimmed)?.value ?: return null
        return match.toDoubleOrNull() // parity:allow stdlib parse; substring false positive
    }
}

/** Diagnostics + registry identifiers for the surface (P1/S11). */
object CommandInputDialogRegistration {
    /** Stable surface id. */
    const val ID: String = "command-input-dialog"

    /** Diagnostics surface slug emitted with the `view.opened` event (P1/S11). */
    const val SLUG: String = "CommandInputDialog"
}

/**
 * The one PII-safe diagnostic this surface emits (P1/S11). Carries only the surface
 * [CommandInputDialogRegistration.SLUG] — never a field value, the typed PIN, or the command — so a diagnostics
 * line can never leak what the operator is entering. Kept free of Compose so it is unit-tested with a recording
 * [Logger]; the composable calls it from its first-composition effect.
 */
object CommandInputDialogDiagnostics {
    private const val VIEW_OPENED = "view.opened"
    private const val SURFACE_KEY = "surface"

    /** Emits the `view.opened` diagnostic for this surface. Call from the composable's first-composition effect. */
    fun recordViewOpened(logger: Logger) {
        logger.info(VIEW_OPENED, mapOf(SURFACE_KEY to CommandInputDialogRegistration.SLUG))
    }
}
